### ISSUE TITLE
Speed up AOCS index build by only scan needed columns.

### ISSUE THREAD

**Issue by wangxuqi** [greenplum-db/gpdb#13006]
## Enhancement
As we know, AOCS will do column projection when we only need to select part columns of a table. This feature can also be used during AOCS table index build to speed up index build when the AOCS table has already contains large data.

The feature only needs a few code change.  Call a new `table_beginscan_strat` that will internal call `aoco_beginscan_extractcolumns`. And what we need to do is just pass a targetList and a qualification to `aoco_beginscan_extractcolumns` to finish projection push down.

**Comment by kainwen** [comment]
@wangxuqi 
Would you please open a PR and do some benchmark test?

Thanks!

**Comment by wangxuqi** [comment]
> you

OK

**Comment by kainwen** [comment]
If we do not have this feature yet, let's keep this issue.

Once the enhancement PR is merged, then let's close this.

**Comment by wangxuqi** [comment]
When I try to implement this feature. I find that it is not simple as I thought. The problem is "block directory".  Because the block directory is lazy created. If we only scan columns we needed during index build. It may leads to a problem that "block directory" lost the block infos of unscanned columns. And after Index build, Although we insert correct index item into index page, but we lost  block directory entries.  As you can see below:

```
create table aocs_index_cols (id int4, a text, b text) with (appendonly=true, orientation=column);
insert into aocs_index_cols values (1, 'foo', 'bar');
create index on aocs_index_cols (id);
set enable_seqscan=off;
```
```
SELECT * FROM bt_page_items('aocs_index_cols_id_idx', 1);
 itemoffset |     ctid     | itemlen | nulls | vars |          data
------------+--------------+---------+-------+------+-------------------------
          1 | (33554432,2) |      16 | f     | f    | 01 00 00 00 00 00 00 00
```
```
select ctid, * from aocs_index_cols where id = 1;
     ctid     | id |  a  |  b
--------------+----+-----+-----
 (33554432,2) |  1 | foo | bar
(1 row)
```
When we only query indexed column, the result is right.

```
select id from aocs_index_cols where id = 1;
 id
----
  1
(1 row)
```
But if we query more columns, the result is wrong.
```
select id, a from aocs_index_cols where id = 1;
 id | a
----+---
(0 rows)
```

```
select * from pg_aoseg.pg_aoblkdir_16569 limit 10;
 segno | columngroup_no | first_row_no |                                  minipage
-------+----------------+--------------+----------------------------------------------------------------------------
     1 |              0 |            1 | \x000000000100000000000000010000000000000000000000000000000100000000000000
(1 row)
```
BlockDirectory only contains cloumn no 0 entry, Lost other column infos.

The function `AppendOnlyBlockDirectory_GetEntry` will return false if can not get minipage. So the wrong result return 0 rows.


As above, Sorry for this feature, It can only be implemented if we always write out block directory entries when insert tuple. Consider lazy build  block directory may more useful than faster index build at most time. 

Close this issue for now.

**Comment by ashwinstar** [comment]
In gpdb6 and before, on first index creation (block directory is created) and only that time all columns are scanned. Later any index creation just scans needed columns.
Similar logic can be coded for master branch as well.

I have further points for discussion on this topic which I would initiate gpdb-dev thread next week.

**Comment by wangxuqi** [comment]
> In gpdb6 and before, on first index creation (block directory is created) and only that time all columns are scanned. Later any index creation just scans needed columns. Similar logic can be coded for master branch as well.
> 
> I have further points for discussion on this topic which I would initiate gpdb-dev thread next week.

Good ideal, I will scan all columns when first index creation (RelationGetNumberOfBlocks(blkdir) == 0). and Use this feature during other index creation. I will reopen it and create a new PR. Thanks for your advice.

**Comment by wangxuqi** [comment]
PR #13012
